src/modal-manager.ts

```diff
@@ -0,0 +1,181 @@
+import {
+  LitElement,
+  html,
+  css,
+  customElement,
+  CSSResult,
+  TemplateResult,
+  property,
+  query,
+  PropertyValues,
+} from 'lit-element';
+
+import './modal-template';
+import { ModalTemplate } from './modal-template';
+import { ModalConfig } from './modal-config';
+
+export enum ModalManagerMode {
+  Modal = 'modal',
+  Closed = 'closed'
+}
+
+export interface ModalManagerInterface extends LitElement {
+  /**
+   * Show a modal from a given ModalConfig
+   *
+   * @param config ModalConfig
+   * @param customModalContent TemplateResult | undefined
+   * @param userClosedModalCallback () => void | undefined an optional callback when the modal is closed
+   */
+  showModal(options: {
+    config: ModalConfig;
+    customModalContent?: TemplateResult;
+    userClosedModalCallback?: () => void;
+  }): Promise<void>;
+
+  /**
+   * Close the currently open modal
+   */
+  closeModal(): void;
+}
+
+@customElement('modal-manager')
+export class ModalManager extends LitElement implements ModalManagerInterface {
+  /**
+   * The current mode of the ModalManager
+   *
+   * Current options are `modal` or `closed`
+   *
+   * @type {ModalManagerMode}
+   * @memberof ModalManager
+   */
+  @property({ type: String, reflect: true }) mode: ModalManagerMode = ModalManagerMode.Closed;
```

Comment:
q: what does `reflect` mean?

Reply:
`reflect` causes the property to update the element tag attributes. Normally if a property changes its value inside the component, the tag stays the same. It's a one-way street, the consumer sets it and it's used inside. With reflect, when the property value changes, it updates the tag too. ie:
```
<modal-manager mode='closed'></modal-manager>
...
modalManager.showModal(...) // causes mode to change to `modal`
```
Without `reflect`, the tag would not change. When you enable `reflect`, the tag would become:
```
<modal-manager mode='modal'></modal-manager>
```

The reason we're using `reflect` here is because the modal visibility is largely just controlled by CSS, but the functionality of triggering that open/close is based on internal logic. The consumer of the modal needs to add these styles:
```
  modal-manager {
    position: absolute;
    width: 100%;
    height: 100%;
    z-index: 250;
  }

  modal-manager[mode='closed'] { <--- these are the reflected properties
    display: none;
  }

  modal-manager[mode='modal'] { <--- these are the reflected properties
    display: block;
  }
```

Reply:
thanks for the break down.